main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+      val e = pending.head
+      pending = pending.tail
+
+      e match {
+        case Equation(Term.Var(x), Term.Var(y), _) =>
+          // Case 1: We have found an equation: `x ~ y`.
+          // Construct a singleton substitution `[x -> y]`.
+          val singleton = BoolSubstitution.singleton(x, Term.Var(y))
+
+          // Apply the singleton substitution to the pending equations.
+          pending = singleton(pending)
+
+          // Apply the singleton substitution to the unsolved equations.
+          unsolved = singleton(unsolved)
+
+          // Update the current substitution with the new binding.
+          subst = singleton @@ subst
+
+        case _ =>
+          // Case 2: We have some other type of equation. Add it to the list of unsolved equations.
+          // The invariant ensures that the current substitution has already been applied to the equation.
+          unsolved = e :: unsolved
+      }
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (unsolved.reverse, subst)
+  }
+
+  /**
+    * Assigns variables. Given a unification equation `x ~ t` we can assign `[x -> t]` if `x` does not occur in `t`.
+    *
+    * For example, given the equation system:
+    *
+    * {{{
+    *    x78914 ~ (c1498 ∧ c1500 ∧ c1501)
+    *    x78914 ~ (x78926 ∧ x78923 ∧ x127244)
+    *    x78923 ~ (x127249 ∧ x127247 ∧ x127248)
+    *    x78926 ~ (x127254 ∧ x127252)
+    * }}}
+    *
+    * We compute the substitution:
+    *
+    * {{{
+    *     x78926 -> (x127254 ∧ x127252)
+    *     x78914 -> (c1498 ∧ c1500 ∧ c1501)
+    *     x78923 -> (x127249 ∧ x127247 ∧ x127248)
+    * }}}
+    *
+    * and we have the remaining equations:
+    *
+    * {{{
+    *     (c1498 ∧ c1500 ∧ c1501) ~ (x127248 ∧ x127244 ∧ x127254 ∧ x127252 ∧ x127249 ∧ x127247)
+    * }}}
+    */
+  private def varAssignment(l: List[Equation], subst0: BoolSubstitution): (List[Equation], BoolSubstitution) = {
+    var subst = subst0
+    var pending = l
+
+    var unsolved: List[Equation] = Nil
+
+    // INVARIANT: The current substitution has been applied to BOTH the unsolved AND pending equations.
+    while (pending.nonEmpty) {
+      // Extract the next equation and update the pending equations.
+      val e = pending.head
+      pending = pending.tail
+
+      e match {
+        case Equation(Term.Var(x), rhs, _) if !rhs.freeVars.contains(x) =>
+          // Case 1: We have found an equation: `x ~ t` where `x` is not free in `t`.
+          // Construct a singleton substitution `[x -> y]`.
+          val singleton = BoolSubstitution.singleton(x, rhs)
+
+          // Apply the singleton substitution to the unsolved equations.
+          pending = singleton(pending)
+
+          // Apply the singleton substitution to the unsolved equations.
+          unsolved = singleton(unsolved)
+
+          // Compose the new substitution with the current substitution.
+          subst = singleton @@ subst
```

Comment:
I think the new substitution whould be to the right?
```
x ~ y + True
y ~ z + False
```
singleton on the left gives `[y -> z + False, x -> y + False]` while singleton on the right gives `[x -> z  + False + True, y -> z + False]`

Reply:
nevermind, i read @@ the wrong way round